dll/win32/shell32/wine/control.c

```diff
@@ -967,65 +960,199 @@ static	void	Control_DoLaunch(CPanel* panel, HWND hWnd, LPCWSTR wszCmd)
             sp = 0;
         }
 
-#ifdef __REACTOS__
-        bActivated = (applet->hActCtx != INVALID_HANDLE_VALUE ? ActivateActCtx(applet->hActCtx, &cookie) : FALSE);
+        if (!applet->proc(applet->hWnd, CPL_STARTWPARMSW, sp, (LPARAM)extraPmts))
+            applet->proc(applet->hWnd, CPL_DBLCLK, sp, applet->info[sp].data);
+
+        Control_UnloadApplet(applet);
+    }
 
-        aCPLPath = GlobalFindAtomW(applet->cmd);
-        if (!aCPLPath)
+    HeapFree(GetProcessHeap(), 0, buffer);
+#else
+    LPWSTR	buffer;
+    LPWSTR       ptr;
+    signed 	sp = -1;
+    LPCWSTR	extraPmts = L"";
+    BOOL        quoted = FALSE;
+    CPlApplet *applet;
+    LPCWSTR wszFirstCommaPosition = NULL, wszSecondCommaPosition = NULL;
+    LPCWSTR wszLastUnquotedSpacePosition = NULL;
+    LPWSTR wszDialogBoxName;
+    int i = 0;
+    SIZE_T nLen = lstrlenW(wszCmd);
+
+    buffer = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*buffer) * (nLen + 1));
+    wszDialogBoxName = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*wszDialogBoxName) * (nLen + 1));
+    if (wszDialogBoxName == NULL || buffer == NULL)
+    {
+        if (buffer != NULL)
+            HeapFree(GetProcessHeap(), 0, buffer);
+        if (wszDialogBoxName != NULL)
+            HeapFree(GetProcessHeap(), 0, wszDialogBoxName);
+        return;
+    }
+
+    /* Search for unquoted commas and spaces. */
+    for (i = 0; i < nLen; i++)
+    {
+        if (wszCmd[i] == '"')
+            quoted = !quoted;
+        if (wszCmd[i] == ',' && !quoted)
+        {
+            if (wszFirstCommaPosition == NULL)
+                wszFirstCommaPosition = &wszCmd[i];
+            else if (wszSecondCommaPosition == NULL)
+                wszSecondCommaPosition = &wszCmd[i];
+        }
+        if (wszCmd[i] == ' ' && !quoted)
         {
-            aCPLPath = GlobalAddAtomW(applet->cmd);
+            wszLastUnquotedSpacePosition = &wszCmd[i];
         }
+    }
 
-        aCPLName = GlobalFindAtomW(L"CPLName");
-        if (!aCPLName)
+    /* If no unquoted commas are found, parameters are either space separated, or the entire string
+     * is a CPL path. */
+    if (wszFirstCommaPosition == NULL)
+    {
+        /* An unquoted space was found in the string. Assume the last word is the dialog box
+         * name/number. */
+        if (wszLastUnquotedSpacePosition != NULL)
         {
-            aCPLName = GlobalAddAtomW(L"CPLName");
+            int nSpaces = 0;
+
+            while (wszLastUnquotedSpacePosition[nSpaces] == L' ')
+                nSpaces++;
+
+            StringCchCopyNW(buffer, nLen, wszCmd, wszLastUnquotedSpacePosition - wszCmd);
+            lstrcpyW(wszDialogBoxName, wszLastUnquotedSpacePosition + nSpaces);
```

Comment:
Why using here lstrcpyW, instead of the corresponding string-safe function? + Same question for the other call below.